### PR TITLE
Add on premise microk8s deployment guide

### PR DESCRIPTION
**Pull request by fiedlerNr9**
This PR adds a tutorial which showcases an on-premise deployment of Flyte using microk8s. 

The tutorial covers steps from setting up a on-premise device(raspberry pi) in a local network to submitting workflows from another development machine against the healthy Flyte deployment. To keep the overview on docs, a infra level directory structure was added as well. 

Open To Do's:
- With this directory structure, markdown links for the aws & k3s deployment needs to be adjusted

### CODE REVIEW

File: docs/on-premises/microk8s/01-preparing-raspberry-pi.md [hunk 1]
```diff
@@ -0,0 +1,34 @@
+# Preparing the Raspberry Pi
+
+Feel free to skip this section, if you already have a ubuntu-server powered machine in your homelab. This section will be a brief walk through the process of preparing a Raspberry Pi before installing microk8s.
```
Comment (davidmirror-ops) [inline docs/on-premises/microk8s/01-preparing-raspberry-pi.md]: Ubuntu Server

File: docs/on-premises/microk8s/01-preparing-raspberry-pi.md [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+
+## Pre-requisites
+### Hardware
+- Raspberry Pi 4 Modell B (8 GB RAM recommended)
```
Comment (davidmirror-ops) [inline docs/on-premises/microk8s/01-preparing-raspberry-pi.md]:
```suggestion
- Raspberry Pi 4 Model B (8 GB RAM recommended)
```

File: docs/on-premises/microk8s/01-preparing-raspberry-pi.md [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+- Raspberry Pi Power Supply & Ethernet Cable
+- Micro SD Card & SD Card Reader
+### Software
+- The [Raspberry Pi Imager](https://www.raspberrypi.com/software/)
+
+## Flash the Micro SD Card
+- Download and run the Raspberry Pi Imager, choose your correct Raspberry Pi Device and select the Micro SD card that is connected to your machine
+- For the Operating System I recommend: `Other general-purpose OS > Ubuntu > Ubuntu Server 22.04.3 LTS (64-bit)`
```
Comment (davidmirror-ops) [inline docs/on-premises/microk8s/01-preparing-raspberry-pi.md]:
```suggestion
- For the Operating System use: `Other general-purpose OS > Ubuntu > Ubuntu Server 22.04.3 LTS (64-bit)`
```

File: docs/on-premises/microk8s/02-install-configure-microk8s.md [hunk 4]
```diff
@@ -0,0 +1,27 @@
+# Install and Configure MikroK8s
+
+To deploy kubernetes on your ubuntu powered device install the microk8s snap.
```
Comment (davidmirror-ops) [inline docs/on-premises/microk8s/02-install-configure-microk8s.md]:
```suggestion
To deploy Kubernetes on your Ubuntu-powered device install the microk8s snap.
```

File: docs/on-premises/microk8s/02-install-configure-microk8s.md [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+> This tutorial uses the 1.25 channel version
+``` bash
+sudo snap install microk8s --classic --channel=1.25
+```
+To enable your current user seamless usage of microk8s command, run the following commands:
```
Comment (davidmirror-ops) [inline docs/on-premises/microk8s/02-install-configure-microk8s.md]:
```suggestion
To enable the `microk8s` command for your current user,  run the following commands:
```

File: docs/on-premises/microk8s/02-install-configure-microk8s.md [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+``` bash
+sudo usermod -a -G microk8s $USER
+sudo chown -f -R $USER ~/.kube
+newgrp microk8s
+```
+To validate the microk8s installation you can run the following command to display the current status of microk8s and enabled/disabled addons:
+``` bash
+microk8s status
+```
+If this successfully displays the microk8s status, your are ready to operate via kubectl like shown in the following example:
```
Comment (davidmirror-ops) [inline docs/on-premises/microk8s/02-install-configure-microk8s.md]:
```suggestion
If it successfully displays the microk8s status, you are ready to operate via `kubectl` like shown in the following example:
```

File: docs/on-premises/microk8s/04-install-flyte.md [hunk 7]
```diff
@@ -0,0 +1,46 @@
+# Install Flyte
+To make sure your future Flyte deployment can access kubernetes services via internal dns, make sure to enable the dns addon of microk8s:
```
Comment (davidmirror-ops) [inline docs/on-premises/microk8s/04-install-flyte.md]:
```suggestion
To make sure your future Flyte deployment can access Kubernetes services via internal DNS, make sure to enable the DNS addon of microk8s:
```

File: docs/on-premises/microk8s/04-install-flyte.md [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+``` bash
+microk8s enable dns
+```
+Download the values file:
+``` bash
+curl -sl https://raw.githubusercontent.com/davidmirror-ops/flyte-the-hard-way/main/docs/on-premises/microk8s/local-values.yaml > local-values.yaml.yaml
+```
+> Make sure to adjust the values for the minio & postgres deployment you just submitted
+
+Install the Flyte binary via helm:
```
Comment (davidmirror-ops) [inline docs/on-premises/microk8s/04-install-flyte.md]:
```suggestion
Install the Flyte binary via Helm:
```

File: docs/on-premises/microk8s/05-add-ingress-and-tls.md [hunk 9]
```diff
@@ -0,0 +1,67 @@
+# Adding Ingress & TLS
+
+This section will demonstrate how to add ingress and tls to your flyte deployment in order to access flyte from any device in your local network.
```
Comment (davidmirror-ops) [inline docs/on-premises/microk8s/05-add-ingress-and-tls.md]:
```suggestion
This section will demonstrate how to add Ingress and TLS to your Flyte deployment to access Flyte from any device in your local network.
```

File: docs/on-premises/microk8s/05-add-ingress-and-tls.md [hunk 10]
```diff
@@ -0,0 +1,67 @@
...
+
+First of all, enable the ingress microk8s addon:
```
Comment (davidmirror-ops) [inline docs/on-premises/microk8s/05-add-ingress-and-tls.md]:
```suggestion
First of all, enable the Ingress microk8s addon:
```

File: docs/on-premises/microk8s/05-add-ingress-and-tls.md [hunk 11]
```diff
@@ -0,0 +1,67 @@
...
+``` bash
+microk8s enable ingress
+```
+
+> For simplicity reasons, this tutorial uses self-signed certificates to enable a tls connection.
```
Comment (davidmirror-ops) [inline docs/on-premises/microk8s/05-add-ingress-and-tls.md]:
```suggestion
> For simplicity reasons, this tutorial uses self-signed certificates to enable a TLS connection.
```

File: docs/on-premises/microk8s/05-add-ingress-and-tls.md [hunk 12]
```diff
@@ -0,0 +1,67 @@
...
+
+``` bash
+KEY_FILE=flyte.key
+CERT_FILE=flyte.crt
+CERT_NAME=flytetls
+HOST=flyte.local
+openssl req -x509 -nodes -days 365 -newkey rsa:2048 -keyout ${KEY_FILE} -out ${CERT_FILE} -subj "/CN=${HOST}/O=${HOST}" -addext "subjectAltName = DNS:${HOST}"
+```
+
+Given the certificate and key, create a kubernetes secret in the flyte namespace:
```
Comment (davidmirror-ops) [inline docs/on-premises/microk8s/05-add-ingress-and-tls.md]:
```suggestion
Given the certificate and key, create a Kubernetes secret in the `flyte` namespace:
```

File: docs/on-premises/microk8s/05-add-ingress-and-tls.md [hunk 13]
```diff
@@ -0,0 +1,67 @@
...
+``` bash
+kubectl create secret tls ${CERT_NAME} --key ${KEY_FILE} --cert ${CERT_FILE} -n flyte
+```
+
+To add ingress and tls to your flyte deployment, add the following yaml to your Flyte binary helm chart and update your deployment:
```
Comment (davidmirror-ops) [inline docs/on-premises/microk8s/05-add-ingress-and-tls.md]:
```suggestion
To add Ingress and TLS to your Flyte deployment, add the following block to your Flyte binary Helm chart and update your deployment:
```

File: docs/on-premises/microk8s/05-add-ingress-and-tls.md [hunk 14]
```diff
@@ -0,0 +1,67 @@
...
+        - "{{ .Values.userSettings.hostName }}"
+      secretName: flytetls
+  commonAnnotations:
+    kubernetes.io/ingress.class: public
+  httpAnnotations:
+    nginx.ingress.kubernetes.io/app-root: /console
+  grpcAnnotations:
+    nginx.ingress.kubernetes.io/backend-protocol: GRPC
+```
+To validate your updated deployment, check the ingress resources in the flyte namespace:
```
Comment (davidmirror-ops) [inline docs/on-premises/microk8s/05-add-ingress-and-tls.md]:
```suggestion
To validate your updated deployment, check the Ingress resources in the flyte namespace:
```

**Review by davidmirror-ops — APPROVED** [review]
Just minor changes.
@fiedlerNr9 this is absolutely amazing work. I learned a lot from your process and hopefully will be incorporating some of it on the other guides

Thank you!

**Comment by fiedlerNr9** [comment]
Thank you so much @davidmirror-ops! I incorporated your suggestions and updated some READMEs according to the new repo structure. I also went through all the steps again with a fresh raspberry and added some missing notes. For the future i might add chapters for auth and logs (microk8s observability addon) :)

Are you able to resolve the conflict on this PR? Not sure if i can do it with the permissions i have on this repo

**Pull request merged**